torchft/checkpointing.py

```diff
@@ -40,36 +41,50 @@ class CheckpointServer(Generic[T]):
         state_dict: a callable that returns the state dict to be transferred
     """
 
-    def __init__(self, state_dict: Callable[[], T]) -> None:
+    def __init__(self, state_dict: Callable[[], T], timeout: timedelta) -> None:
         self._checkpoint_lock = threading.Lock()
         self._disallowed = False
         self._step = -1
+        self._timeout = timeout
 
         ckpt_server = self
 
         class RequestHandler(BaseHTTPRequestHandler):
-            def do_GET(self):
-                with ckpt_server._checkpoint_lock:
-                    step = ckpt_server._step
+            # set request socket timeout to avoid hanging forever
+            timeout = self._timeout.total_seconds()
 
-                    if self.path != f"/checkpoint/{step}":
-                        self.send_response(400)
-                        self.send_header("Content-type", "text/plain")
+            def do_GET(self):
+                try:
+                    # validate socket timeout is actually set
+                    assert self.connection.gettimeout() == self.timeout
+
+                    with ckpt_server._checkpoint_lock:
+                        step = ckpt_server._step
+
+                        if self.path != f"/checkpoint/{step}":
+                            self.send_response(400)
+                            self.send_header("Content-type", "text/plain")
+                            self.end_headers()
+                            self.err(
+                                f"invalid checkpoint requested, serving {step} but got {self.path}"
+                            )
+                            return
+
+                        self.send_response(200)
+                        self.send_header(
+                            "Content-type", "tensor"
```

Comment:
application/octet-stream?

Reply:
just wrapped this in a try catch but changed